builtin/providers/aws/resource_aws_s3_bucket_object.go

```diff
@@ -0,0 +1,117 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/hashicorp/terraform/helper/schema"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+func resourceAwsS3BucketObject() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsS3BucketObjectPut,
+		Read:   resourceAwsS3BucketObjectRead,
+		Update: resourceAwsS3BucketObjectPut,
+		Delete: resourceAwsS3BucketObjectDelete,
+
+		Schema: map[string]*schema.Schema{
+			"bucket": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"key": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"source": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"etag": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsS3BucketObjectPut(d *schema.ResourceData, meta interface{}) error {
+	s3conn := meta.(*AWSClient).s3conn
+
+	bucket := d.Get("bucket").(string)
+	key := d.Get("key").(string)
+	source := d.Get("source").(string)
+
+	file, err := os.Open(source)
+
+	if err != nil {
+		return fmt.Errorf("Error opening S3 bucket object source (%s): %s", source, err)
+	}
+
+	resp, err := s3conn.PutObject(
+		&s3.PutObjectInput{
+			Bucket: aws.String(bucket),
+			Key:    aws.String(key),
+			Body:   file,
+		})
+
+	if err != nil {
+		return fmt.Errorf("Error putting object in S3 bucket (%s): %s", bucket, err)
+	}
+
+	d.Set("etag", resp.ETag)
+	d.SetId(key)
+	return nil
+}
+
+func resourceAwsS3BucketObjectRead(d *schema.ResourceData, meta interface{}) error {
+	s3conn := meta.(*AWSClient).s3conn
+
+	bucket := d.Get("bucket").(string)
+	key := d.Get("key").(string)
+	etag := d.Get("etag").(string)
+
+	resp, err := s3conn.HeadObject(
+		&s3.HeadObjectInput{
+			Bucket:  aws.String(bucket),
+			Key:     aws.String(key),
+			IfMatch: aws.String(etag),
+		})
+
+	if err != nil {
+		// if there is an error reading the object we assume it's not there.
+		d.SetId("")
+		log.Printf("Error Reading Object (%s): %s", key, err)
+		return nil
+	}
```

Comment:
I just tried to scope this, but with no luck:

```go
log.Printf("\n\n--- Error: %#v\n------\n", err)
if awsErr, ok := err.(awserr.Error); ok {
	log.Printf("\n\n-----\n AWS ERror here:\n\tCode: %s\n\tMessage: %s", awsErr.Code(), awsErr.Message())

	log.Printf("\n\n-----\n main error: %s", err.Error())
}
```

Response:

```
: --- Error: &awserr.requestError{awsError:(*awserr.baseError)(0xc208348b10), statusCode:404, requestID:""}
: ------
: 2015/08/03 10:06:44
:
: -----
:  AWS ERror here:
: 	Code:
: 	Message:
```

:scream: 

@phinze I did check `GetObject` (instead of `HeadObject`), which does return `NoSuchKey` for the code. I'm not sure if we should switch though, because of the included network traffic for potentially larger objects. Thoughts?